language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,346 @@
+/// The Vote module is used to allow voting on proposals on the chain.
+/// It is typically not going to be used directly, but is intended to be
+/// used as a library for modules which want to perform operations which
+/// require m-of-n approvals from accounts on chain.
+/// A typical workflow would look like the following
+/// * Module M creates a ballot with a given `Proposal` and an approval policy using `create_ballot`
+/// * It receives the BallotID corresponding to the ballot
+/// * It submits votes using the `vote` function from the voters
+/// * If a vote causes a ballot to be approved, `vote` returns `true` and Module M can proceed with the operation requested by the `Proposal`
+module DiemFramework::Vote {
+
+    use Std::BCS;
+    use Std::Errors;
+    use Std::Event;
+    use Std::Signer;
+    use Std::Vector;
+    use DiemFramework::DiemTimestamp;
+
+    /// An unique identifier for a ballot. A counter is stored
+    /// under each proposers address which is incremented
+    /// every time a new ballot is created. The proposers
+    /// address is also part of the ballot id.
+    struct BallotID has store, copy, drop {
+        counter: u64,
+        proposer: address,
+    }
+
+    /// Ballot is a struct which contains a Proposal on which
+    /// votes are gathered. A ballot is started by a proposer
+    /// and it carries the proposal, the approval policy,
+    /// start/expiration timestamps
+    struct Ballot<Proposal: store + drop> has store, copy, drop {
+        /// Details for the proposal being voted on.
+        proposal: Proposal,
+        /// A human readable type for the proposal - ex: "create_validator_owner",
+        /// "freeze_account", "create_vasp", etc.
+        /// This lives outside the `proposalProposal` to make it easy for
+        /// indexers to index the ballots which have been proposed and
+        /// categorize them into "types"
+        proposal_type: vector<u8>,
+        /// The num_votes_required for this proposal to be approved
+        num_votes_required: u64,
+        /// A vector of addresses which are allowed to vote on this ballot.
+        /// Duplicates are allowed. Duplicates represent the "weight" of a voter
+        /// We represent an address using the bcs serialized bytes of an `address`
+        allowed_voters: vector<vector<u8>>,
```

Comment:
So I did this in my previous PR [here](https://github.com/diem/diem/pull/9160/files#diff-94f28b17c42f30270e231430fabefaaf05253856e6a6d17c797b828cbf901d4e), but did not find it useful since it was just a struct with two functions : `get_allowed_voters` and `get_num_votes_required` with no other logic. I can reintroduce it here as well.

Reply:
I was thinking about a more abstract trait. In Rust nomenclature it'd be something like
```
trait ApprovalPolicy {
  fn add_vote(...) -> bool
}
```
We could have some basic implementation like
```
struct ThresholdVoting {
  voters: vector<WeightedVoter>,
  threshold: u64,
  votes_received: vector<WeightedVoter>,
}
```
The implementations would keep both the rules, as well as the votes.

Reply:
I was thinking of doing it this way. The Ballot struct in the Vote module stores the threshold, allowed_voters, votes_received.
```
module DiemFramework::Vote {
struct Ballot {
    ...
    num_votes_required: u64,
    allowed_voters: vector<WeightedVoter>,
    votes_received: vector<WeightedVoter>,
    ...
}
}
```

And each of the modules using the vote define their own policy which they store themselves

```
module DiemFramework::CreateValidator {

struct CreateValidatorOwnerPolicy has key {
    num_votes_required: u64,
    allowed_voters: vector<Vote::WeightedVoter>,
}    

struct CreateValidatorOperatorPolicy has key {
    num_votes_required: u64,
    allowed_voters: vector<Vote::WeightedVoter>,
}
    
}
```